include/PetriEngine/Colored/VariableVisitor.h

```diff
@@ -205,24 +205,51 @@ namespace PetriEngine {
 
             virtual void accept(const AddExpression* add)
             {
+                bool lastIsVarExpr = false;
                 for (const auto& elem : *add) {
-                    for(auto& pair : _varModifiers){
-                        std::unordered_map<uint32_t, int32_t> newMap;
-                        pair.second.push_back(newMap);
+                    for (auto& pair : _varModifiers) {
+                        pair.second.emplace_back();
                     }
+
                     _index = 0;
+                    size_t sizeBefore = _varModifiers.size();
                     elem->visit(*this);
+                    size_t sizeAfter = _varModifiers.size();
+
+                    lastIsVarExpr = sizeBefore != sizeAfter;
+                }
+
+                if (lastIsVarExpr) {
+                    for (auto& pair : _varModifiers) {
+                        pair.second.emplace_back();
+                    }
                 }
             }
 
             virtual void accept(const SubtractExpression* sub)
             {
+                for (auto& pair : _varModifiers) {
+                    pair.second.emplace_back();
+                }
                 _index = 0;
                 (*sub)[0]->visit(*this);
                 //We ignore the restrictions imposed by the subtraction for now
                 if(_include_subtracts){
+                    for (auto& pair : _varModifiers) {
+                        pair.second.emplace_back();
+                    }
                     _index = 0;
+                    size_t sizeBefore = _varModifiers.size();
                     (*sub)[1]->visit(*this);
+                    size_t sizeAfter = _varModifiers.size();
+
+                    bool lastIsVarExpr = sizeBefore != sizeAfter;
```

Comment:
What happens if we have multiple var-exressions at the end? Will this still work?

Reply:
This is fixed in the latest PR.